metalearn/metafeatures/base.py

```diff
@@ -0,0 +1,146 @@
+        them.
+    argmap
+        A custom map of ``computer``'s argument names to the global resource names that will be passed as
+        ``computer``'s arguments when ``computer`` is called.
+    """
+
+    def __init__(
+        self, computer: Callable, returns: List[str], argmap: Optional[Dict[str,Any]] = None
+    ) -> None:
+        argspec = inspect.getfullargspec(computer)
+        # TODO: If needed, add support for `computer` functions that use these types of arguments.
+        if (
+            argspec.varargs is not None or argspec.varkw is not None or argspec.defaults is not None or
+            len(argspec.kwonlyargs) > 0
+        ):
+            raise ValueError('`computer` must use only positional arguments with no default values')
+
+        self.computer = computer
+        self.returns = returns
+        self.argmap = {arg_name: arg_name for arg_name in argspec.args}
+
+        if argmap is not None:
+            # override computer arg value with developer provided values
+            # Note each value in `argmap` is a global resource name (e.g. `"XSample"`) or a literal value (e.g. `5`)
+            self.argmap.update(argmap)
+        
+    
+    def __call__(self, *args, **kwargs):
+        """
+        Allows a ``ResourceComputer`` instance to be callable. Just forwards all arguments on to self.computer.
+        """
+        return self.computer(*args, **kwargs)
+
+    @property
+    def name(self) -> str:
+        """Returns the function name of self.computer"""
+        return self.computer.__name__
+
+
+class MetafeatureComputer(ResourceComputer):
+
+    def __init__(
+        self,
+        computer: Callable,
+        returns: List[str], # TODO: Add support for passing just a string, not a list?
+        problem_type: ProblemType,
+        groups: List[MetafeatureGroup],
+        argmap: Optional[Dict[str,str]] = {}
+    ) -> None:
+        """
+        Decorates ``computer``, a metafeature computing function
+        with metadata about that function.
+        
+        Parameters
+        ----------
+        computer
+            The function that computes the metafeatures.
+        returns
+            The names of the metafeatures that ``computer`` returns, specified in
+            the same order as ``computer`` returns them.
+        problem_type
+            The type of ML problem `computer`'s metafeatures can be computed for.
+        groups
+            The metafeature groups this computer's returned metafeatures belong to.
+            e.g. statistical, info-theoretic, simple, etc.
+        argmap
+            A custom map of ``computer``'s argument names to the global resource names
+            that will be passed as ``computer``'s arguments when ``computer`` is called.
+        """
+        super(MetafeatureComputer, self).__init__(computer, returns, argmap)
+        self.groups = groups
+        self.problem_type = problem_type
+
+
+class ResourceComputerMap(MutableMapping):
```

Comment:
what is the difference between using `MutableMapping` and just `dict`?

Reply:
**TLDR;**
`collections.abc.MutableMapping` is much simpler to subclass than `dict`. Since `dict`'s methods don't compose up from each other, you have to define a lot more methods on your subclass to implement all the dictionary functionality with `dict` than you do with `MutableMapping`. In short, the developers of Python have made `MutableMapping` available as the preferred way to implement a class with dictionary functionality. `dict` is geared more towards usage than sub-classing.

**The Long Answer**
http://www.kr41.net/2016/03-23-dont_inherit_python_builtin_dict_type.html. See also https://treyhunner.com/2019/04/why-you-shouldnt-inherit-from-list-and-dict-in-python/

Reply:
Great reasons!